src/Core/src/Response.php

```diff
@@ -85,39 +111,130 @@ public function resolve(?float $timeout = null): bool
+        $indexMap = [];
+        $httpResponses = [];
+        $httpClient = null;
+        foreach ($responses as $index => $response) {
+            if (null !== $response->resolveResult && (!$downloadBody || $response->bodyDownloaded)) {
+                yield $index => $response;
+
+                continue;
+            }
+
+            if (null === $httpClient) {
+                $httpClient = $response->httpClient;
+            }
+            $httpResponses[] = $response->httpResponse;
+            $indexMap[$hash = \spl_object_id($response->httpResponse)] = $index;
+            $responseMap[$hash] = $response;
         }
 
-        if (400 <= $statusCode) {
-            throw $this->resolveResult = new ClientException($this->httpResponse);
+        // no response provided (or all responses already resolved)
+        if (empty($httpResponses)) {
+            return;
         }
 
-        if (300 <= $statusCode) {
-            throw $this->resolveResult = new RedirectionException($this->httpResponse);
+        if (null === $httpClient) {
+            throw new InvalidArgument('At least one response should have contain an Http Client');
         }
 
-        return $this->resolveResult = true;
+        foreach ($httpClient->stream($httpResponses, $timeout) as $httpResponse => $chunk) {
+            $hash = \spl_object_id($httpResponse);
+            try {
+                if ($chunk->isTimeout()) {
+                    // Receiving a timeout mean all responses are inactive.
+                    break;
+                }
+            } catch (TransportException $e) {
+                unset($indexMap[$hash]);
+                if (empty($indexMap)) {
+                    // early exit if all statusCode are known. We don't have to wait for all responses
+                    return;
+                }
+
+                continue;
+            }
+
+            $response = $responseMap[$hash] ?? null;
+            // Check if null, just in case symfony yield an unexpected response.
+            if (null === $response) {
+                continue;
+            }
+
+            if (!$response->streamStarted && '' !== $chunk->getContent()) {
+                $response->streamStarted = true;
+            }
+
+            // index could be null if already yield
+            $index = $indexMap[$hash] ?? null;
+            if ($chunk->isLast()) {
+                $response->bodyDownloaded = true;
+                if (null !== $index && $downloadBody) {
+                    unset($indexMap[$hash]);
+                    yield $index => $response;
+                }
+            }
+            if ($chunk->isFirst()) {
+                try {
+                    // call handleStatus to set internal state: `resolveResult` and ignore errors
+                    $response->handleStatus();
+                } catch (Exception $e) {
+                    // unset resolveResult because of refcount: the exception contains a reference to the `$response`
```

Comment:
this one was very hard to find!!! see https://3v4l.org/GI9rU

When an object contains an exception created by a static method, it creates a circular reference that prevent the __destructor being called

Reply:
Oh. wow. I had no idea. 

How did you find it?

Reply:
Added a test to assert that exception is triggered on destruct => test failed

Spent 4 hours with php meminfo, dump, to understand what happens.

Note: I refactored this code. Instead of unset, I store the "components" of the exception in an array and instanciate it when calling "resolve"